attestation/jsonattestor/json.go

```diff
@@ -0,0 +1,213 @@
+			"",
+			func(a attestation.Attestor, sqStr string) (attestation.Attestor, error) {
+				att, ok := a.(*Attestor)
+				if !ok {
+					return a, fmt.Errorf("unexpected attestor type: %T", a)
+				}
+				sqMap, err := ParseSubjectQueries(sqStr)
+				if err != nil {
+					return a, fmt.Errorf("parseSubjectQueries: %w", err)
+				}
+				att.SubjectQueries = sqMap
+				return att, nil
+			},
+		),
+	)
+}
+
+func New() *Attestor {
+	return &Attestor{
+		subjects:       make(map[string]cryptoutil.DigestSet),
+		SubjectQueries: make(map[string]string),
+	}
+}
+
+func ParseSubjectQueries(sqStr string) (map[string]string, error) {
+	out := make(map[string]string)
+	s := strings.TrimSpace(sqStr)
+	if s == "" {
+		return out, nil
+	}
+	pairs := strings.Split(s, ";")
+	for _, pair := range pairs {
+		pair = strings.TrimSpace(pair)
+		if pair == "" {
+			continue
+		}
+		parts := strings.SplitN(pair, "=", 2)
+		if len(parts) != 2 {
+			return nil, fmt.Errorf("invalid subject-queries pair %q (missing '=')", pair)
+		}
+		name := strings.TrimSpace(parts[0])
+		expr := strings.TrimSpace(parts[1])
+		if name == "" || expr == "" {
+			return nil, fmt.Errorf("invalid subject-queries pair %q (empty name or expression)", pair)
+		}
+		out[name] = expr
+	}
+	return out, nil
+}
+
+func (a *Attestor) Name() string {
+	return Name
+}
+
+func (a *Attestor) Type() string {
+	return Type
+}
+
+func (a *Attestor) RunType() attestation.RunType {
+	return RunType
+}
+
+func (a *Attestor) Schema() *jsonschema.Schema {
+	return jsonschema.Reflect(a)
+}
+
+func (a *Attestor) Attest(ctx *attestation.AttestationContext) error {
+	prods := ctx.Products()
+	if len(prods) == 0 {
+		return errors.New("no products to attest")
+	}
+	path := pickOneJSONFile(prods)
+	if path == "" {
+		return errors.New("no .json or .yaml/.yml file found among products")
+	}
+	fullPath := filepath.Join(ctx.WorkingDir(), path)
+	orig, err := os.ReadFile(fullPath)
+	if err != nil {
+		return fmt.Errorf("read error: %w", err)
+	}
+
+	// original-file subject
+	dsOrig, err := cryptoutil.CalculateDigestSetFromBytes(orig, ctx.Hashes())
+	if err == nil && len(dsOrig) > 0 {
+		a.subjects["original-file"] = dsOrig
+	}
+
+	maybeJSON, e := yaml.YAMLToJSON(orig)
+	if e != nil {
+		maybeJSON = orig
+	}
+	var any interface{}
+	if e := json.Unmarshal(maybeJSON, &any); e != nil {
+		return fmt.Errorf("invalid JSON/YAML: %w", e)
+	}
+
+	a.CanonicalJSON, err = jsoncanonicalizer.Transform(maybeJSON)
+	if err != nil {
+		return fmt.Errorf("canonicalization error: %w", err)
+	}
+	if dsCan, e := cryptoutil.CalculateDigestSetFromBytes(a.CanonicalJSON, ctx.Hashes()); e == nil && len(dsCan) > 0 {
+		a.subjects["canonical-json"] = dsCan
+	}
+
+	var root interface{}
+	if e := json.Unmarshal(a.CanonicalJSON, &root); e != nil {
+		return fmt.Errorf("re-unmarshal canonical JSON: %w", e)
+	}
+	for subjName, jqExpr := range a.SubjectQueries {
+		runOneJQ(a.subjects, subjName, jqExpr, root, ctx)
```

Comment:
Do you think we should log here possible errors due to malformed input?

(nit/future improvement): maybe runOneJQ could use async (goroutine) for parallel multiple jsons?

Reply:
done